File: HotelAuthService/HotelAuthService/Services/HotelService.cs
using HotelAuthService.DTOs;
using HotelAuthService.Repositories;

namespace HotelAuthService.Services
{
    public class HotelService : IHotelService
    {
        private readonly IHotelRepository _hotelRepository;

        public HotelService(IHotelRepository hotelRepository)
        {
            _hotelRepository = hotelRepository;
        }

        public async Task<IEnumerable<HotelDto>> GetAllHotelsAsync()
        {
            var hotels = await _hotelRepository.GetAllHotelsAsync();

            return hotels.Select(h => new HotelDto
            {
                HotelId = h.HotelId,
                Name = h.Name,
                Location = h.Location,
                Description = h.Description,
                Amenities = h.Amenities,
                Rating = h.Rating,
                Rooms = h.Rooms.Select(r => new RoomDto
                {
                    RoomId = r.RoomId,
                    RoomNumber = r.RoomNumber,
                    Type = r.Type,
                    Description = r.Description,
                    PricePerNight = r.PricePerNight,
                    MaxGuests = r.MaxGuests,
                    IsAvailable = r.IsAvailable,
                    HotelId = r.HotelId
                }).ToList()
            });
        }

        public async Task<IEnumerable<HotelDto>> SearchHotelsAsync(
            string location)
        {
            var hotels = await _hotelRepository
                .SearchHotelsAsync(location);

            return hotels.Select(h => new HotelDto
            {
                HotelId = h.HotelId,
                Name = h.Name,
                Location = h.Location,
                Description = h.Description,
                Amenities = h.Amenities,
                Rating = h.Rating,
                Rooms = h.Rooms.Select(r => new RoomDto
                {
                    RoomId = r.RoomId,
                    RoomNumber = r.RoomNumber,
                    Type = r.Type,
                    Description = r.Description,
                    PricePerNight = r.PricePerNight,
                    MaxGuests = r.MaxGuests,
                    IsAvailable = r.IsAvailable,
                    HotelId = r.HotelId
                }).ToList()
            });
        }

        public async Task<HotelDto> GetHotelByIdAsync(int hotelId)
        {
            var hotel = await _hotelRepository
                .GetHotelByIdAsync(hotelId);

            if (hotel == null)
            {
                throw new Exception("Hotel not found");
            }

            return new HotelDto
            {
                HotelId = hotel.HotelId,
                Name = hotel.Name,
                Location = hotel.Location,
                Description = hotel.Description,
                Amenities = hotel.Amenities,
                Rating = hotel.Rating,
                Rooms = hotel.Rooms.Select(r => new RoomDto
                {
                    RoomId = r.RoomId,
                    RoomNumber = r.RoomNumber,
                    Type = r.Type,
                    Description = r.Description,
                    PricePerNight = r.PricePerNight,
                    MaxGuests = r.MaxGuests,
                    IsAvailable = r.IsAvailable,
                    HotelId = r.HotelId
                }).ToList()
            };
        }

        public async Task<IEnumerable<RoomDto>> GetRoomsByHotelIdAsync(
            int hotelId)
        {
            var rooms = await _hotelRepository
                .GetRoomsByHotelIdAsync(hotelId);

            return rooms.Select(r => new RoomDto
            {
                RoomId = r.RoomId,
                RoomNumber = r.RoomNumber,
                Type = r.Type,
                Description = r.Description,
                PricePerNight = r.PricePerNight,
                MaxGuests = r.MaxGuests,
                IsAvailable = r.IsAvailable,
                HotelId = r.HotelId
            });
        }

        public async Task<RoomDto> GetRoomByIdAsync(int roomId)
        {
            var room = await _hotelRepository
                .GetRoomByIdAsync(roomId);

            if (room == null)
            {
                throw new Exception("Room not found");
            }

            return new RoomDto
            {
                RoomId = room.RoomId,
                RoomNumber = room.RoomNumber,
                Type = room.Type,
                Description = room.Description,
                PricePerNight = room.PricePerNight,
                MaxGuests = room.MaxGuests,
                IsAvailable = room.IsAvailable,
                HotelId = room.HotelId
            };
        }

        public async Task<IEnumerable<HotelDto>> SearchHotelsAsync(
    string? location,
    decimal? minPrice,
    decimal? maxPrice,
    string? amenity)
        {
            var hotels = await _hotelRepository
                .SearchHotelsAsync(location, minPrice, maxPrice, amenity);

            return hotels.Select(h => new HotelDto
            {
                HotelId = h.HotelId,
                Name = h.Name,
                Location = h.Location,
                Description = h.Description,
                Amenities = h.Amenities,
                Rating = h.Rating,
                Rooms = h.Rooms.Select(r => new RoomDto
                {
                    RoomId = r.RoomId,
                    RoomNumber = r.RoomNumber,
                    Type = r.Type,
                    Description = r.Description,
                    PricePerNight = r.PricePerNight,
                    MaxGuests = r.MaxGuests,
                    IsAvailable = r.IsAvailable,
                    HotelId = r.HotelId
                }).ToList()
            });
        }
    }
}

File: frontend/HotelBookingApp/src/app/core/services/hotel.service.ts
import { Injectable } from '@angular/core';
import { HttpClient, HttpParams } from '@angular/common/http';
import { Observable } from 'rxjs';

export interface RoomDto {
  roomId: number;
  roomNumber: string;
  type: string;
  description: string;
  pricePerNight: number;
  maxGuests: number;
  isAvailable: boolean;
  hotelId: number;
}

export interface HotelDto {
  hotelId: number;
  name: string;
  location: string;
  description: string;
  amenities: string;
  rating: number;
  rooms: RoomDto[];
}

@Injectable({
  providedIn: 'root'
})
export class HotelService {

  private apiUrl = 'https://localhost:7055/api/hotel';

  constructor(private http: HttpClient) {}

  getAllHotels(): Observable<HotelDto[]> {
    return this.http.get<HotelDto[]>(`${this.apiUrl}/all`);
  }

  searchHotels(
    location?: string,
    minPrice?: number,
    maxPrice?: number,
    amenity?: string
  ): Observable<HotelDto[]> {
    let params = new HttpParams();

    if (location) params = params.set('location', location);
    if (minPrice) params = params.set('minPrice', minPrice.toString());
    if (maxPrice) params = params.set('maxPrice', maxPrice.toString());
    if (amenity) params = params.set('amenity', amenity);

    return this.http.get<HotelDto[]>(`${this.apiUrl}/search`, { params });
  }

  getHotelById(hotelId: number): Observable<HotelDto> {
    return this.http.get<HotelDto>(`${this.apiUrl}/${hotelId}`);
  }

  getRoomsByHotelId(hotelId: number): Observable<RoomDto[]> {
    return this.http.get<RoomDto[]>(`${this.apiUrl}/${hotelId}/rooms`);
  }

  getRoomById(roomId: number): Observable<RoomDto> {
    return this.http.get<RoomDto>(`${this.apiUrl}/room/${roomId}`);
  }
}